### PR TITLE
Bugfix and testing for unconstrained optimization

### LINKED ISSUE #157: PSQP and CONMIN not working for unconstrained problems
## Type of issue
What types of issue is it?

Bug
## Description
wrappers for CONMIN and PSQP pass incorrectly formatted variables to FORTRAN code for unconstrained problems.
In addition, it appears that both methods also need the "dummy Constraint" as implemented in SLSQP as the code tries to access a constraint variable that doesn't exit (which the introduction of the dummy constraint fixes).

I attached modified code for pyCONMIN.py and pyPSQP.py that fix the problems
### Steps to reproduce issue
1. Run attached test script implementing Rosenbrock's function (modified from pyOpt test script)
2. …
3. …

### Current behavior
Python throws an error message when calling the FORTRAN wrapper. Once that's fixed, non-existent variable "gcon" causes another error (fixed with the introduction of the dummy constraint)

### Expected behavior
no error

### Code version (if relevant)
**Python version:** 3.7

[bugfix.zip](https://github.com/mdolab/pyoptsparse/files/5198120/bugfix.zip)

### PR DESCRIPTION
**Pull request by marcomangano**
## Purpose
As pointed out in #157, few optimizers in pyOptSparse do not natively support an unconstrained optimization problem formulation.
We fix this issue using `dummyConstraint` option from the `Optimization()` class, indeed providing a dummy constraint to the algorithm to complete the input set.
As this kind of optimization problem was untested, the option was not turned on for PSQP, CONMIN, and NLPQLP.

Now the feature is enabled, supporting unconstrained optimization problems throughout pyOptSparse. 
Moreover, I added a new test based on the Rosenbrock function to verify that all our optimizers support unconstrained problems.

Closes #157 and partially addresses #145.

## Type of change

- Bugfix (non-breaking change which fixes an issue)

## Testing
Run the entire test suite or just the newly added `test_rosenbrock.py`

## Checklist

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

File: test/test_rosenbrock.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        # Check Solution
+        self.fStar1 = 0.0
+
+        self.xStar1 = np.ones(n)
+
+        dv = sol.getDVs()
+        sol_xvars = [sol.variables["xvars"][i].value for i in range(n)]
+
+        assert_allclose(sol_xvars, dv["xvars"], atol=tol, rtol=tol)
+        # we check either optimum via try/except
```
Comment (nwu63) [inline test/test_rosenbrock.py:106]: Please remove these comments

File: test/test_rosenbrock.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        # try:
+        assert_allclose(sol.objectives["obj"].value, self.fStar1, atol=tol, rtol=tol)
+        assert_allclose(dv["xvars"], self.xStar1, atol=tol, rtol=tol)
+        # except AssertionError:
```
Comment (nwu63) [inline test/test_rosenbrock.py:110]: and of course these

**Review by nwu63 — COMMENTED** [review]
The rest looks good. I've verified that the test you added fails for the current `master` branch, for CONMIN, PSQP and NLPQLP. Good work

File: test/test_rosenbrock.py (hunk 2 above)
Comment (marcomangano) [reply test/test_rosenbrock.py:110]: Yes thanks, I missed them. I copied and adapted `test_hs071.py` for this test. I removed the try/except statement but not the comments

**Review by nwu63 — APPROVED** [review]
Yep looks good. But you can see why I wanted to refactor the tests, there is now significant code duplication between the different tests, so we should pull these out into a util package or something like that.

**Comment by marcomangano** [comment]
Agreed, probably we can move the different objective functions under a separate class / method and have a single set of functions that call pyOptSparse from a single file. I haven't checked the other tests, but I believe `hs_071` and `hs_015` have minimal differences, and of course in this new test I changed a few hardcoded values and commented out all the calls that refer to the `"con"` keyword.

File: test/test_rosenbrock.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        metadata = hist.getMetadata()
+        self.assertEqual(metadata["optimizer"], optimizer)
+        metadata_def_keys = ["optName", "optOptions", "nprocs", "startTime", "endTime", "optTime", "version"]
+        for key in metadata_def_keys:
+            self.assertIn(key, metadata)
+        hist.getOptProb()
+
+        # Info checks
+        self.assertEqual(hist.getDVNames(), ["xvars"])
+        # self.assertEqual(hist.getConNames(), ["con"])
```
Comment (sseraj) [inline test/test_rosenbrock.py:125]: Do we want to keep these lines as comments or remove them?

**Review by sseraj — COMMENTED** [review]
I have also checked that the relevant tests fail without the patch.

File: test/test_rosenbrock.py (hunk 3 above)
Comment (marcomangano) [reply test/test_rosenbrock.py:125]: These come from `test_hs_071` as well. I don't have any strong opinion about that , I left them as "placeholders" for a later refactoring, but we can take them off, up to you guys!

Comment (sseraj) [reply test/test_rosenbrock.py:125]: @nwu63 Any thoughts on this?

Comment (nwu63) [reply test/test_rosenbrock.py:125]: I think we should remove these

**Review by sseraj — APPROVED** [review]

**Review by nwu63 — APPROVED** [review]
@marcomangano it's 1am, go sleep!

**Pull request merged**